### PR TITLE
ttl: introduce `ttl.PhysicalTable` to provide some common methods for TTL

### LINKED ISSUE #39428: introduce `ttl.PhysicalTable` to provide some common methods for TTL

### PR DESCRIPTION
**Pull request by lcwangchao**
### What problem does this PR solve?

Issue Number: close #39428

Problem Summary:

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: ttl/session.go [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+// RunInTxn executes the specified function in a txn
+func (s *Session) RunInTxn(ctx context.Context, fn func() error) (err error) {
+	if _, err = s.ExecuteSQL(ctx, "BEGIN"); err != nil {
+		return err
+	}
+
+	success := false
+	defer func() {
+		if !success {
+			_, err = s.ExecuteSQL(ctx, "ROLLBACK")
```
Comment (bb7133) [inline ttl/session.go:76]: Should we consider 'COMMIT'?

File: ttl/table.go [hunk 2]
```diff
@@ -15,27 +15,141 @@
...
+func (t *PhysicalTable) EvalExpireTime(ctx context.Context, se *Session, now time.Time) (expire time.Time, err error) {
+	tz := se.GetSessionVars().TimeZone
+
+	expireExpr := t.TTLInfo.IntervalExprStr
+	unit := ast.TimeUnitType(t.TTLInfo.IntervalTimeUnit)
+
+	var rows []chunk.Row
+	rows, err = se.ExecuteSQL(
+		ctx,
+		// FROM_UNIXTIME does not support value, so we use `FROM_UNIXTIME(0) + INTERVAL <current_ts>` to present current time
```
Comment (bb7133) [inline ttl/table.go:145]: "FROM_UNIXTIME does not support value"

What does this mean?

File: ttl/session.go (hunk 1 above)
Comment (lcwangchao) [reply ttl/session.go:76]: Yes.. It's mistake, I'll add more tests.

File: ttl/table.go (hunk 2 above)
Comment (lcwangchao) [reply ttl/table.go:145]: FROM_UNIXTIME does not support negative value. The comment lost the word "negative" ...

File: ttl/session.go [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+
+// RunInTxn executes the specified function in a txn
+func (s *Session) RunInTxn(ctx context.Context, fn func() error) (err error) {
+	if _, err = s.ExecuteSQL(ctx, "BEGIN"); err != nil {
+		return err
+	}
+
+	success := false
+	defer func() {
+		if !success {
```
Comment (bb7133) [inline ttl/session.go:75]: It's better to change it to: `if success {`, but this change is not required.

**Review by bb7133 — APPROVED** [review]
LGTM

**Review by hawkingrei — APPROVED** [review]

**Review by YangKeao — APPROVED** [review]
A little question :thinking:

File: ttl/table.go [hunk 4]
```diff
@@ -15,27 +15,141 @@
...
+	tz := se.GetSessionVars().TimeZone
+
+	expireExpr := t.TTLInfo.IntervalExprStr
+	unit := ast.TimeUnitType(t.TTLInfo.IntervalTimeUnit)
+
+	var rows []chunk.Row
+	rows, err = se.ExecuteSQL(
+		ctx,
+		// FROM_UNIXTIME does not support negative value, so we use `FROM_UNIXTIME(0) + INTERVAL <current_ts>` to present current time
+		fmt.Sprintf("SELECT FROM_UNIXTIME(0) + INTERVAL %d SECOND - INTERVAL %s %s", now.Unix(), expireExpr, unit.String()),
```
[the remark this answers is not in the document]
Comment (lcwangchao) [reply ttl/table.go:146]: The difference is that the `now` is an argument of the function for easy to mock in unit test and `SELECT NOW() ...` is a hard code to use the real time

**Pull request merged**